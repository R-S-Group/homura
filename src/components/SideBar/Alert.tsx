import React from 'react';
import { CloseIcon } from '../Icon';

export const Alert: React.FC<{ onClick: () => void }> = ({ onClick }) => (
  <div
    className="text-xs bg-red-100 border border-red-400 text-red-700 px-4 py-3 rounded relative"
    role="alert"
  >
    <strong className="font-bold">Search RSS failed!</strong>
    <span className="block">
      Please check the link and the internet connection.
    </span>
    <span className="absolute top-0 bottom-0 right-0 px-4 py-3">
      <div
        tabIndex={0}
        role="button"
        className="h-4 w-4 text-red-700 cursor-pointer"
        onClick={onClick}
        onKeyDown={onClick}
      >
        <CloseIcon />
      </div>
    </span>
  </div>
);
