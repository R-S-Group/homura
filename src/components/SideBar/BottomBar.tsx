import React from 'react';
import { PlusIcon, RefreshIcon } from '../Icon';
import { IconContainer } from '../LogoIcon';

export enum Step {
  EnterUrl,
  EnterName,
}

export type BottomBarProps = {
  step: Step | null;
  onClickPlus: () => void;
  onClickSync: () => void;
  render: (step: Step) => JSX.Element;
  loading: boolean;
};

export const BottomBar: React.FC<BottomBarProps> = ({
  step,
  onClickPlus,
  onClickSync,
  render,
  loading,
}) => (
  <div role="toolbar" className="relative flex justify-between text-gray-300">
    {step !== null && (
      <div className="absolute bottom-0 inset-x-0 z-10">{render(step)}</div>
    )}
    <IconContainer onClick={onClickPlus}>
      <PlusIcon />
    </IconContainer>
    <IconContainer onClick={onClickSync}>
      <RefreshIcon spin={loading} />
    </IconContainer>
  </div>
);
