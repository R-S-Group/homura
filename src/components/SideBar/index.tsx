import React from 'react';

export type SideBarProps = {
  overview: JSX.Element;
  bottom: JSX.Element;
  onClick: () => void;
};

export const SideBar: React.FC<SideBarProps> = ({
  overview,
  bottom,
  children,
  onClick,
}) => {
  return (
    <div
      data-testid="source-side-bar"
      role="presentation"
      onClick={onClick}
      className="w-1/6 bg-gray-700 flex flex-col h-screen select-none"
    >
      <div className="flex-1 thin-scroll relative">
        {overview}
        <div className="pt-10" />
        {children}
      </div>
      {bottom}
    </div>
  );
};
