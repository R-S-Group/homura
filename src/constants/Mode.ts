export enum Mode {
    All,
    Unread,
    Starred,
}
