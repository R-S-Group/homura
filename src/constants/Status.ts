export enum Status {
  Idle,
  Pending,
  Succeeded,
  Failed,
}
