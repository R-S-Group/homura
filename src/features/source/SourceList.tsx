import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { SideBar } from '../../components/SideBar';
import { Header } from '../../components/SideBar/Header';
import { SideBarItem } from '../../components/SideBar/SideBarItem';
import { AppDispatch } from '../../store';
import {
  fetchSources,
  selectSourceList,
  setCurrentSource,
  syncSources,
  unsubscribeById,
  updateSourceById,
} from './sourceSlice';

export const SourceList: React.FC<{ bottom: JSX.Element }> = ({ bottom }) => {
  const dispatch = useDispatch<AppDispatch>();
  const { list, activeId, mode, totalCount } = useSelector(selectSourceList);
  const isFirstRender = React.useRef(true);

  React.useEffect(() => {
    if (isFirstRender.current) {
      dispatch(syncSources(mode));
      isFirstRender.current = false;
      return;
    }
    dispatch(fetchSources(mode));
  }, [dispatch, mode]);

  const overview = (
    <Header
      mode={mode}
      active={activeId}
      count={totalCount}
      onClick={(id) => dispatch(setCurrentSource(id))}
    />
  );

  return (
    <SideBar
      onClick={() => dispatch(setCurrentSource(null))}
      overview={overview}
      bottom={bottom}
    >
      {list.map(({ id, name, count, icon }) => (
        <SideBarItem
          key={id.toString()}
          id={id}
          url={icon}
          name={name}
          count={count}
          active={id === activeId}
          onConfirmModify={(nextName) =>
            dispatch(updateSourceById({ id, name: nextName }))
          }
          onUnsubscribe={() => dispatch(unsubscribeById(id))}
          onClick={() => dispatch(setCurrentSource(id))}
        />
      ))}
    </SideBar>
  );
};
