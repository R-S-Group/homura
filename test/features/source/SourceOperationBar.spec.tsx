import { configureStore } from '@reduxjs/toolkit';
import { fireEvent, getByLabelText, getByRole, render, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { Step } from '../../../src/components/SideBar/BottomBar';
import { Mode } from '../../../src/constants/Mode';
import { Status } from '../../../src/constants/Status';
import * as modeSlice from '../../../src/features/mode/modeSlice';
import { SourceOperationBar } from '../../../src/features/source/SourceOperationBar';
import * as sourceSlice from '../../../src/features/source/sourceSlice';

function setup(
    preloadedState: {
        source: Partial<sourceSlice.State>;
        mode?: Mode;
    } = {
        source: {},
        mode: Mode.All,
    },
) {
    const initialSourceState = {
        list: [],
        activeId: undefined,
        fetchListStatus: Status.Idle,
        syncListStatus: Status.Idle,
        subscribeLink: '',
        subscribeName: '',
        subscribeStep: undefined,
        subscribeStatus: Status.Idle,
        subscribeError: undefined,
    };
    const state: {
        source: sourceSlice.State;
        mode: Mode;
    } = {
        source: { ...initialSourceState, ...preloadedState.source },
        mode: preloadedState.mode ?? Mode.All,
    };
    const store = configureStore({
        reducer: { source: sourceSlice.sourceReducer, mode: modeSlice.modeReducer },
        preloadedState: state,
    });

    render(
        <Provider store={store}>
            <SourceOperationBar />
        </Provider>,
    );

    return {
        store,
    };
}

describe('SourceOperationBar component', () => {
    it('should match snapshot when not rendered', () => {
        setup();

        expect(screen.getByRole('toolbar')).toMatchSnapshot();
    });

    it('should match snapshot enterUrlStep', () => {
        setup({
            source: {
                subscribeStep: Step.EnterUrl,
            },
        });

        expect(screen.getByRole('toolbar')).toMatchSnapshot();
    });

    it('should match snapshot enterNameStep', () => {
        setup({
            source: {
                subscribeStep: Step.EnterName,
            },
        });

        expect(screen.getByRole('toolbar')).toMatchSnapshot();
    });

    it('should click button', () => {
        const subscribeToSourceSpy = jest.spyOn(sourceSlice, 'subscribeToSource');
        setup({
            source: {
                subscribeStep: Step.EnterName,
            },
            mode: Mode.Starred,
        });

        const subscribeContent = screen.getByTestId('sidebar-subscribe-bar-content');
        const input = getByLabelText(subscribeContent, 'subscribed-source-name');
        const button = getByRole(subscribeContent, 'button');
        fireEvent.change(input, { target: { value: 'test name' } });
        fireEvent.click(button);
        expect(subscribeToSourceSpy).toBeCalledWith({ mode: Mode.Starred, name: 'test name' });
    });
});
