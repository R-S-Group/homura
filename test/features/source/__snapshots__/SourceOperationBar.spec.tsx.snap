// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceOperationBar component should match snapshot enterNameStep 1`] = `
<div
  class="sidebar-bottom-bar"
  role="toolbar"
>
  <div
    class="sidebar-bottom-bar__step"
  >
    <div
      class="sidebar-subscribe-bar__link"
    />
    <div
      class="sidebar-subscribe-bar__content"
      data-testid="sidebar-subscribe-bar-content"
    >
      <div
        class="flex-1"
      >
        <input
          aria-label="subscribed-source-name"
          class="sidebar-subscribe-bar__input"
          placeholder="Save as Name"
          value=""
        />
      </div>
      <div
        class="p-2  sidebar-subscribe-bar__icon"
        role="button"
        tabindex="0"
      >
        <div
          class="w-5 h-5 m-auto  "
        >
          <svg
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.707-9.293a1 1 0 00-1.414-1.414L9 10.586 7.707 9.293a1 1 0 00-1.414 1.414l2 2a1 1 0 001.414 0l4-4z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    class="p-2  "
    role="button"
    tabindex="0"
  >
    <div
      class="w-5 h-5 m-auto  "
    >
      <svg
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 6v6m0 0v6m0-6h6m-6 0H6"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </div>
  <div
    class="p-2  "
    role="button"
    tabindex="0"
  >
    <div
      class="w-5 h-5 m-auto  "
    >
      <svg
        class=""
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 4v5h.582m15.356 2A8.001 8.001 0 004.582 9m0 0H9m11 11v-5h-.581m0 0a8.003 8.003 0 01-15.357-2m15.357 2H15"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`SourceOperationBar component should match snapshot enterUrlStep 1`] = `
<div
  class="sidebar-bottom-bar"
  role="toolbar"
>
  <div
    class="sidebar-bottom-bar__step"
  >
    <div
      class="sidebar-feed-search-bar__container"
    >
      <div
        class="flex-1"
      >
        <input
          class="sidebar-feed-search-bar__input"
          placeholder="Feed URL"
          value=""
        />
      </div>
      <div
        class="p-2  sidebar-feed-search-bar__icon show"
        role="button"
        tabindex="0"
      >
        <div
          class="w-5 h-5 m-auto  "
        >
          <svg
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 9a2 2 0 114 0 2 2 0 01-4 0z"
            />
            <path
              clip-rule="evenodd"
              d="M10 18a8 8 0 100-16 8 8 0 000 16zm1-13a4 4 0 00-3.446 6.032l-2.261 2.26a1 1 0 101.414 1.415l2.261-2.261A4 4 0 1011 5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    class="p-2  "
    role="button"
    tabindex="0"
  >
    <div
      class="w-5 h-5 m-auto  "
    >
      <svg
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 6v6m0 0v6m0-6h6m-6 0H6"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </div>
  <div
    class="p-2  "
    role="button"
    tabindex="0"
  >
    <div
      class="w-5 h-5 m-auto  "
    >
      <svg
        class=""
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 4v5h.582m15.356 2A8.001 8.001 0 004.582 9m0 0H9m11 11v-5h-.581m0 0a8.003 8.003 0 01-15.357-2m15.357 2H15"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`SourceOperationBar component should match snapshot when not rendered 1`] = `
<div
  class="sidebar-bottom-bar"
  role="toolbar"
>
  <div
    class="p-2  "
    role="button"
    tabindex="0"
  >
    <div
      class="w-5 h-5 m-auto  "
    >
      <svg
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 6v6m0 0v6m0-6h6m-6 0H6"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </div>
  <div
    class="p-2  "
    role="button"
    tabindex="0"
  >
    <div
      class="w-5 h-5 m-auto  "
    >
      <svg
        class=""
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 4v5h.582m15.356 2A8.001 8.001 0 004.582 9m0 0H9m11 11v-5h-.581m0 0a8.003 8.003 0 01-15.357-2m15.357 2H15"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </div>
</div>
`;
